pyro/distributions/transforms/affine_coupling.py

```diff
@@ -125,11 +128,86 @@ def log_abs_det_jacobian(self, x, y):
-            _, log_scale = self.hypernet(x1)
+            _, log_scale = self.nn(x1)
             log_scale = clamp_preserve_gradients(log_scale, self.log_scale_min_clip, self.log_scale_max_clip)
         return log_scale.sum(-1)
 
 
+@copy_docs_from(ConditionalTransformModule)
+class ConditionalAffineCoupling(ConditionalTransformModule):
+    """
+    An implementation of the affine coupling layer of RealNVP (Dinh et al., 2017) that conditions on an additional
+    context variable and uses the bijective transform,
+
+        :math:`\\mathbf{y}_{1:d} = \\mathbf{x}_{1:d}`
+        :math:`\\mathbf{y}_{(d+1):D} = \\mu + \\sigma\\odot\\mathbf{x}_{(d+1):D}`
+
+    where :math:`\\mathbf{x}` are the inputs, :math:`\\mathbf{y}` are the outputs, e.g. :math:`\\mathbf{x}_{1:d}`
+    represents the first :math:`d` elements of the inputs, and :math:`\\mu,\\sigma` are shift and translation
+    parameters calculated as the output of a function inputting only :math:`\\mathbf{x}_{1:d}`.
+
+    That is, the first :math:`d` components remain unchanged, and the subsequent :math:`D-d` are shifted and
+    translated by a function of the previous components.
+
+    Together with :class:`~pyro.distributions.ConditionalTransformedDistribution` this provides a way to create
+    richer variational approximations.
+
+    Example usage:
+
+    >>> from pyro.nn import DenseNN
+    >>> input_dim = 10
+    >>> split_dim = 6
+    >>> base_dist = dist.Normal(torch.zeros(input_dim), torch.ones(input_dim))
+    >>> hypernet = DenseNN(split_dim, [10*input_dim], [input_dim-split_dim, input_dim-split_dim])
+    >>> transform = AffineCoupling(split_dim, hypernet)
+    >>> pyro.module("my_transform", transform)  # doctest: +SKIP
+    >>> flow_dist = dist.TransformedDistribution(base_dist, [transform])
+    >>> flow_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse of the Bijector is required when, e.g., scoring the log density of a sample with
+    :class:`~pyro.distributions.ConditionalTransformedDistribution`. This implementation caches the inverse of the
+    Bijector when its forward operation is called, e.g., when sampling from
+    :class:`~pyro.distributions.TransformedDistribution`. However, if the cached value isn't available, either because
+    it was overwritten during sampling a new value or an arbitary value is being scored, it will calculate it manually.
+
+    This is an operation that scales as O(1), i.e. constant in the input dimension. So in general, it is cheap
+    to sample *and* score (an arbitrary value) from :class:`~pyro.distributions.transforms.ConditionalAffineCoupling`.
+
+    :param split_dim: Zero-indexed dimension :math:`d` upon which to perform input/output split for transformation.
+    :type split_dim: int
+    :param nn: A neural network whose forward call returns a real-valued mean and logit-scale as a tuple. The input
```

Comment:
Maybe specify that the nn constructor must take in a `context` argument to supplement the input dims?

Reply:
Done!